examples/s_expression.rs

```diff
@@ -0,0 +1,331 @@
+/// Remember from before that Atoms form one half of our language.
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum Atom {
+  Num(i32),
+  Keyword(String),
+  Boolean(bool),
+  BuiltIn(BuiltIn),
+}
+
+/// The remaining half is Lists. We implement these as recursive Expressions.
+/// For a list of numbers, we have `'(1 2 3)`, which we'll parse to:
+/// ```
+/// Expr::Quote(vec![Expr::Constant(Atom::Num(1)),
+///                  Expr::Constant(Atom::Num(2)),
+///                  Expr::Constant(Atom::Num(3))])
+/// Quote takes an S-expression and prevents evaluation of it, making it a data
+/// structure that we can deal with programmatically. Thus any valid expression
+/// is also a valid data structure in Lisp itself.
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum Expr {
+  Constant(Atom),
+  /// (func-name arg1 arg2)
+  Application(Box<Expr>, Vec<Expr>),
+  /// (if predicate do-this)
+  If(Box<Expr>, Box<Expr>),
+  /// (if predicate do-this otherwise-do-this)
+  IfElse(Box<Expr>, Box<Expr>, Box<Expr>),
+  /// '(3 (if (+ 3 3) 4 5) 7)
+  Quote(Vec<Expr>),
+}
+
+/// Continuing the trend of starting from the simplest piece and building up,
+/// we start by creating a parser for the built-in functions.
+fn parse_builtin<'a, E: ParseError<&'a str>>(i: &'a str) -> IResult<&'a str, BuiltIn, E> {
+  // alt gives us the result of first parser that succeeds, of the series of
+  // parsers we give it
+  alt((
+    // map lets us process the parsed output, in this case we know what we parsed,
+    // so we ignore the input and return the Builtin directly
+    map(tag("+"), |_| BuiltIn::Plus),
+    map(tag("-"), |_| BuiltIn::Minus),
+    map(tag("*"), |_| BuiltIn::Times),
+    map(tag("/"), |_| BuiltIn::Divide),
+    map(tag("="), |_| BuiltIn::Equal),
+    map(tag("not"), |_| BuiltIn::Not),
+  ))(i)
+}
+
+/// Our boolean values are also constant, so we can do it the same way
+fn parse_bool<'a, E: ParseError<&'a str>>(i: &'a str) -> IResult<&'a str, Atom, E> {
+  alt((map(tag("#t"), |_| Atom::Boolean(true)), map(tag("#f"), |_| Atom::Boolean(false))))(i)
+}
+
+/// The next easiest thing to parse are keywords.
+/// We introduce some error handling combinators: `context` for human readable errors
+/// and `cut` to prevent back-tracking.
+///
+/// Put plainly: `preceded(tag(":"), cut(alpha1))` means that once we see the `:`
+/// character, we have to see one or more alphabetic chararcters or the input is invalid.
+fn parse_keyword<'a, E: ParseError<&'a str>>(i: &'a str) -> IResult<&'a str, Atom, E> {
+  map(context("keyword", preceded(tag(":"), cut(alpha1))), |sym_str: &str| {
+    Atom::Keyword(sym_str.to_string())
+  })(i)
+}
+
+/// Next up is number parsing. We're keeping it simple here by accepting any number (> 1)
+/// of digits but ending the program if it doesn't fit into an i32.
+fn parse_num<'a, E: ParseError<&'a str>>(i: &'a str) -> IResult<&'a str, Atom, E> {
+  alt((
+    map(digit1, |digit_str: &str| Atom::Num(digit_str.parse::<i32>().unwrap())),
```

Comment:
you can use `map_res` to avoir that unwrap: `map_res(digit1, |digit_str: &str| digit_str.parse::<i32>().map(Atom::Num))`